lib/ramble/ramble/util/command_runner.py

```diff
@@ -0,0 +1,173 @@
+# <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+# option. This file may not be copied, modified, or distributed
+# except according to those terms.
+from typing import List
+from spack.util.executable import CommandNotFoundError, ProcessError
+from ramble.util.executable import which
+from ramble.util.logger import logger
+
+
+class CommandRunner:
+    """Runner for executing external commands
+
+    This class provides a generic wrapper on external commands, to provide a
+    unified way to handle dry-run execution of external commands.
+
+    Can be inherited to construct custom command runners.
+    """
+
+    def __init__(self, name=None, command=None, shell="bash", dry_run=False):
+        """
+        Ensure required command is found in the path
+        """
+        self.name = name
+        self.command = None
+        self.dry_run = dry_run
+        if command is not None:
+            try:
+                self.command = which(command, required=True)
+            except CommandNotFoundError:
+                raise RunnerError(f"Command {name} is not found in path")
+
+        self.shell = shell
+
+    def get_version(self):
+        """Hook to get the version of the executable
+
+        Should return a string representation of the executable's version.
+        """
+        pass
+
+    def set_dry_run(self, dry_run=False):
+        """
+        Set the dry_run state of this spack runner
+        """
+        self.dry_run = dry_run
+
+    def execute(self, executable, args: List[str], return_output: bool = False):
+        """Wrapper around execution of a command
+
+        Handles execution of a command when the execution path is dependent on
+        whether dry run is enabled or disabled.
+
+        Args:
+            executable (spack.util.executable.Executable): Executable to run with arguments
+            args (list(str)): List of string arguments to pass into executable
+            return_output (bool): Whether the output of the command should be returned or not
+        """
+        if not self.dry_run:
+            return self._run_command(executable, args, return_output=return_output)
+        else:
+            return self._dry_run_print(executable, args, return_output=return_output)
+
+    def _raise_validation_error(self, command, validation_type):
+        """Wrapper to raise a validation error for this command"""
+        raise ValidationFailedError(
+            f'Validation of: "{self.name} {command}" failed '
+            f' with a validation_type of "{validation_type}"'
+        )
+
+    def _dry_run_print(self, executable, args, return_output=False):
+        """Print the command that would be executed if dry-run was false.
+
+        Args match the execute method.
+        """
+        logger.msg(f"DRY-RUN: would run {executable}")
+        logger.msg(f"         with args: {args}")
+        if return_output:
+            return None
```

Comment:
Given the implicit None return, this feels a bit redundant?

Reply:
Good point. Fixed.